tests/unit/TestNetRemoteDataStreamingReactors.cxx

```diff
@@ -70,3 +70,62 @@ DataStreamWriter::NextWrite()
         StartWritesDone();
     }
 }
+
+DataStreamReader::DataStreamReader(NetRemoteDataStreaming::Stub* client, DataStreamDownloadRequest* request)
+{
+    client->async()->DataStreamDownload(&m_clientContext, request, this);
+    StartCall();
+    StartRead(&m_data);
+}
+
+void
+DataStreamReader::OnReadDone(bool isOk)
+{
+    if (isOk) {
+        m_numberOfDataBlocksReceived++;
+        StartRead(&m_data);
+    }
+    // If read fails, then there is likely no more data to be read, so do nothing.
+}
+
+void
+DataStreamReader::OnDone(const grpc::Status& status)
+{
+    std::unique_lock lock(m_readStatusGate);
+
+    m_status = status;
+    m_done = true;
+    m_readsDone.notify_one();
+}
+
+grpc::Status
+DataStreamReader::Await(uint32_t* numberOfDataBlocksReceived, DataStreamOperationStatus* operationStatus)
+{
+    std::unique_lock lock(m_readStatusGate);
+    static constexpr auto timeoutValue = 10s;
+
+    const auto isDone = m_readsDone.wait_for(lock, timeoutValue, [this] {
+        return m_done;
+    });
+
+    // Handle timeout from waiting for reads to be completed.
+    if (!isDone) {
+        DataStreamOperationStatus status{};
+        status.set_code(DataStreamOperationStatusCode::DataStreamOperationStatusCodeTimedOut);
+        status.set_message("Timeout occurred while waiting for all reads to be completed");
+        *m_data.mutable_status() = std::move(status);
+    }
+
+    // Handle mismatched sequence number and number of data blocks received.
+    if (m_data.sequencenumber() != m_numberOfDataBlocksReceived) {
+        DataStreamOperationStatus status{};
+        status.set_code(DataStreamOperationStatusCode::DataStreamOperationStatusCodeFailed);
```

Comment:
Looking at this now, would it be helpful to know when mismatched sequencing occurs? Might this allow us to help quantify the number of blocks dropped/lost?

Reply:
That's a good idea. It would be easy to tell - basically move this check into `OnReadDone`. We'd need to decide if we want to stop reading and cancel the RPC as soon as the first sequencing mismatch occurs, or if we want to keep going and later count how many blocks were dropped.

Reply:
Going to defer this until a bit later since it's test code - tracked by #186